### PR TITLE
Replace Log with stdlib logger

### LINKED ISSUE #127: Corral should use the ponyc standard library logger
The existing code is almost a duplicate in terms of functionality. There's no need for a bespoke solution.

### PR DESCRIPTION
**Pull request by rkallos**
Closes #127

### CODE REVIEW

**Comment by rkallos** [comment]
Upon further reading, should I change the loggers to all print to a single fd to also fix #121?

File: corral/util/log.pony [hunk 1]
```diff
@@ -99,16 +19,27 @@ primitive CodeLogFormatter is LogFormatter
...
-primitive LevelLogFormatter is LogFormatter
-  fun apply(level: LogLevel, msg: String, loc: SourceLoc): String =>
-    (recover String(level.string().size()
-      + msg.size()
-      + 4)
-    end)
-     .> append(level.string())
-     .> append(": ")
-     .> append(msg)
+primitive LevelLogFormatter
```
Comment (SeanTAllen) [inline corral/util/log.pony:22]: Where is this used? I'm wondering how valuable it is at the point.

File: corral/util/log.pony [hunk 2]
```diff
@@ -1,87 +1,7 @@
...
-    log(LvlWarn, msg, loc)
-
-  fun err(msg: String, loc: SourceLoc = __loc) =>
-    log(LvlErrr, msg, loc)
-
-interface val LogFormatter
-  fun apply(level: LogLevel, msg: String, loc: SourceLoc): String
+use "logger"
 
 primitive CodeLogFormatter is LogFormatter
```
Comment (SeanTAllen) [inline corral/util/log.pony:3]: Where is this used? I'm wondering how valuable it is at the point.

File: corral/util/log.pony [hunk 3]
```diff
@@ -1,87 +1,7 @@
...
-  fun apply(level: LogLevel, msg: String, loc: SourceLoc): String =>
```
Comment (SeanTAllen) [inline corral/util/log.pony:84]: I assume the log level where was so it would be polymorphic. Did you see that there's no need for being able to switch between these formatters?

File: corral/util/log.pony [hunk 4]
```diff
@@ -99,16 +19,27 @@ primitive CodeLogFormatter is LogFormatter
...
+            + 4)
+          end)
+          .> append(level_string)
+          .> append(": ")
+          .> append(msg)
+      end
+    end
 
 primitive SimpleLogFormatter is LogFormatter
-  fun apply(level: LogLevel, msg: String, loc: SourceLoc): String =>
```
Comment (SeanTAllen) [inline corral/util/log.pony:113]: I assume the log level where was so it would be polymorphic. Did you see that there's no need for being able to switch between these formatters?

File: corral/util/log.pony [hunk 5]
```diff
@@ -99,16 +19,27 @@ primitive CodeLogFormatter is LogFormatter
...
-    (recover String(level.string().size()
-      + msg.size()
-      + 4)
-    end)
-     .> append(level.string())
-     .> append(": ")
-     .> append(msg)
+primitive LevelLogFormatter
+  fun apply(level: LogLevel): LogFormatter =>
```
Comment (SeanTAllen) [inline corral/util/log.pony:23]: this seems odd. it only takes a log level?

How is this being used? The lambda concerns me. I'm not sure why that is needed.

File: corral/util/error.pony [hunk 6]
```diff
@@ -1,4 +0,0 @@
-class val Error
```
Comment (SeanTAllen) [inline corral/util/error.pony:1]: was this not used?

Comment (rkallos) [reply corral/util/error.pony:1]: I removed it because it clashed with logger.Error, and it only ever held strings to be passed to a logger.

**Comment by SeanTAllen** [comment]
> Upon further reading, should I change the loggers to all print to a single fd to also fix #121?

No. That's more of a "let's look into the interleaving and make sure its not a bug".

File: corral/util/error.pony (hunk 6 above)
Comment (SeanTAllen) [reply corral/util/error.pony:1]: Makes sense.

File: corral/util/log.pony (hunk 5 above)
Comment (rkallos) [reply corral/util/log.pony:23]: If I understand the original code properly, the previous LevelLogFormatter prepended the log type (fine, info, warn, error) to the message being logged. It is only used in main.pony, either as a one-off before exiting or after parsing the command line arguments.

If you like, I could get rid of it by changing the call sites to add the log message type.

Comment (SeanTAllen) [reply corral/util/log.pony:23]: Ah so the existing pony standard library logger doesn't have the log level. That makes sense. So, this wraps every log call later and appends the level. That makes sense.

Can you add some doc comments to both the apply method and the primitive to make clear what it is doing?

File: corral/util/log.pony (hunk 3 above)
Comment (rkallos) [reply corral/util/log.pony:84]: Log level was there because log.pony had its own implementation of `LogFormatter` which accepted a log level argument. The log level argument was used in `LevelLogFormatter` to prepend the log message type to the message being logged.

Comment (SeanTAllen) [reply corral/util/log.pony:84]: Yup that makes sense now. I like how you moved the log level to being purely at the top level. 

Some package level comments for the logging subsystem for corral that details how the different parts work together should be added the PR so that folks in the future won't be asking the same questions.

File: corral/bundle/bundle.pony [hunk 7]
```diff
@@ -9,30 +10,30 @@ class Bundle
...
   let deps: Map[String, Dep ref] = deps.create()
   var modified: Bool = false
 
-  new iso create(dir': FilePath, log': Log) =>
+  new iso create(dir': FilePath, log': Logger[String]) =>
     dir = dir'
     log = log'
     info = InfoData(JsonObject)
-    log.info("Created bundle in " + dir.path)
+    log(Info) and log.log("Created bundle in " + dir.path)
```
Comment (SeanTAllen) [inline corral/bundle/bundle.pony:22]: here's a question, should there be a "Log" class for corral that keeps the previous form of different methods and it contains the Logger actor and the check to see whether to send a message? It's certainly cleaner and hides to `log(Info)`.

By hiding that, we could have single comments per method on the Logger class that explains why that is done to short circuit sending a message to the actor.

It might make the code base more approachable for folks who don't really know Pony.

Thoughts?

I'm not sure how that would come together with the other changes. Perhaps it might be problematic. I'm not sure.

Comment (rkallos) [reply corral/bundle/bundle.pony:22]: I looked into that approach, but when I read the docs for the logger package and noticed the use of boolean short-circuiting to avoid the work of building strings, I didn't see a clear way to preserve the encapsulation provided by Log and to get the near zero-cost logging offered by the logger API. That's why I chose to replace all the call sites with the idiomatic use of logger.

Comment (SeanTAllen) [reply corral/bundle/bundle.pony:22]: Makes sense.

Comment (SeanTAllen) [reply corral/bundle/bundle.pony:22]: It's amusing that you are reminding me of this given that I wrote those docs for the logger package after writing the logger package.

😆

**Comment by SeanTAllen** [comment]
This looks good aside from needing some documentation for the corral logging package and a couple of the "trickier" formatter usages.

**Comment by SeanTAllen** [comment]
@rkallos this will need to be rebased as it now has conflicts with changes in cmd_update.pony

**Comment by rkallos** [comment]
Rebased :smiley:

**Comment by SeanTAllen** [comment]
Looks like you have compilation errors after the rebase.

**Comment by rkallos** [comment]
Note to self: Run tests locally after rebasing.

**Comment by SeanTAllen** [comment]
@rkallos ping me after you've had time to add the documentation/comments and I'll have another look.

File: corral/util/log.pony [hunk 8]
```diff
@@ -99,16 +19,35 @@ primitive CodeLogFormatter is LogFormatter
...
-    (recover String(level.string().size()
-      + msg.size()
-      + 4)
-    end)
-     .> append(level.string())
-     .> append(": ")
-     .> append(msg)
+primitive LevelLogFormatter
+  """
+  LevelLogFormatter produces LogFormatters that prepend a log level to entries.
```
Comment (SeanTAllen) [inline corral/util/log.pony:24]: This doesn't really feel like explains the purpose.

We want to have the log level in all the output for filtering, searching etc that folks might want to do with an unstructured log.

Rather than adding the format to each string at all call locations, we create one of these in main that in turn wraps the other formatters appending the log-level"

That's more than the comment for this primitive, but part of that should be included in this primitive.

I'd like to see log.pony moved from corral/util to corral/log or corral/logging and have a package level doc as well that explains the relationship between this logger and the others and how logging is in general used in corral (as detailed above).

Including the discussion about the short circuit that your commit switched to.

Basically cover everything that I felt the need to ask you about as part oft this PR.

Comment (rkallos) [reply corral/util/log.pony]: Thank you for the clarification. I've added some more clarification. Let me know what you think.

File: corral/log/log.pony [hunk 9]
```diff
@@ -1,29 +1,40 @@
...
-     .> append(": ")
-     .> append(msg)
+Logging in corral uses the 'logger' package from the standard library, with
+custom implementations of the 'LogFormatter' interface.
+
+The API for 'logger' makes use of boolean short-circuiting in order to avoid the
+work of building strings that won't get logged. This can make log statements
+somewhat verbose:
+
+```
```
Comment (SeanTAllen) [inline corral/log/log.pony:11]: can you pony codefence type?

Comment (rkallos) [reply corral/log/log.pony]: Done. It's `pony`, not `ponylang`, right?

File: corral/log/log.pony [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  filter.
+
+  LevelLogFormatter avoids the need to prepend the log level at all logging call
+  sites by creating a special-purpose LogFormatter whose sole purpose is to
+  prepend the log level.
+
+  LevelLogFormatter is used by the logger(s) created in Main.
+  """
+  fun apply(level: LogLevel): LogFormatter =>
+    """
```
Comment (SeanTAllen) [inline corral/log/log.pony:40]: Lines 33 to 35 should be most of the apply method comment.

Creates lambda to... blah blah blah.

File: corral/log/log.pony [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  """
+  LevelLogFormatter produces LogFormatters that prepend a log level to log
+  messages. Log messages that contain the log level are easier to search and to
+  filter.
+
+  LevelLogFormatter avoids the need to prepend the log level at all logging call
+  sites by creating a special-purpose LogFormatter whose sole purpose is to
+  prepend the log level.
+
+  LevelLogFormatter is used by the logger(s) created in Main.
```
Comment (SeanTAllen) [inline corral/log/log.pony:37]: information from 33 to 37 should be in the package level comments.

File: corral/log/log.pony [hunk 12]
```diff
@@ -0,0 +1,64 @@
+"""
+# Logging in corral
+
```
Comment (SeanTAllen) [inline corral/log/log.pony:3]: I think the most important thing for the package docs is to detail the various components and how they interact so...

uses Pony standard logging.

There's a log level formatter that acts as a wrapper around others so that we get INFO etc output without having to spread throughout the codebase.

And here is how it is used with other formatters.

File: corral/log/log.pony (hunk 9 above)
Comment (SeanTAllen) [reply corral/log/log.pony]: correct

**Comment by SeanTAllen** [comment]
I left some more docs comments. I think what is really missing is, particularly at the package level is...

Written for someone who is coming in to modify logging, add new functionality.

Description of how logging is done and works.

How formatters interact with one another.

What each formatter does.

Basically what you need to be aware of to make changes.

I hope that makes sense. Not sure if it does.

**Comment by SeanTAllen** [comment]
I had to run an errand... that lead to a thought.

For the documentation, there are two audiences.

People who want to enhance/change the logging that exists.
People who are going to add logging to new parts of corral.

The basic informational needs of both groups should be met.

**Comment by rkallos** [comment]
Upon further reflection, I think my approach with `LevelLogFormatter` is incorrect. It prepends the same log level to all log messages, regardless of user intent:

```pony
let logger = StringLogger(Info, env.err, LevelLogFormatter(Info))
logger(Error) and logger.log("some error message")
// logs "INFO: some error message"
```

To fix this, I think I will introduce a new primitive that accepts a log level and a string and does the prepending there. The resulting code should look like this:

```pony
let logger = StringLogger(Info, env.err)
logger(Error) and logger.log(LogMsg(Error, "some error message"))
// logs "ERRR: some error message"
```

This avoids the downside of having `Log` be a sort of wrapper class around `Logger` (not being able to save on work via short-circuiting) while also removing the need for `LevelLogFormatter`.

Alternately, this might make a compelling argument for adding a `log_level` argument to the `LogFormatter` interface in the standard library, though that would require a great deal of care; it seems like it would be a breaking change.

**Comment by SeanTAllen** [comment]
@rkallos This is a good real world use case for the standard library logger. I think it might have highlighted a need for the inclusion of Log Level in the LogFormatter stdlib interface.

Would you be up to opening an RFC to make that change? It's small but we need to run it by people. I'm fully in support of it. 

This PR could then be updated so that it would be a small change to finish off and merge once the standard library has been updated.

**Comment by rkallos** [comment]
[New RFC](https://github.com/ponylang/rfcs/pull/175)

**Comment by rkallos** [comment]
Now that 0.36.0 is out, work on this can continue. It'd probably be best if I started over, as a lot of the details are now lost to me.
